Add ecosystems package API version metrics

d33bs opened the pull request:
<!-- _modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_
_referenced with modifications from [pycytominer](https://github.com/cytomining/pycytominer/blob/main/.github/PULL_REQUEST_TEMPLATE.md)_ -->

# Description

This PR adds metrics from the [Ecosystems packages API](https://packages.ecosyste.ms/) to deepen the detection of version details for the given repository. I tried to keep these generalized and numeric, summarizing the data returned by the Almanack.

Thanks for any feedback!

Closes #160 

## What is the nature of your change?

- [ ] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own contributions.
- [x] I have commented my content, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to related documentation (outside of book content).
- [x] My changes generate no new warnings.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests that prove my additions are effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

falquaddoomi commented on src/almanack/metrics/data.py:
```diff
@@ -770,3 +776,43 @@ def get_github_build_metrics(
     return {}
+
+
+def get_ecosystems_package_metrics(repo_url: str) -> Dict[str, Any]:
```
Correct me if I'm wrong, but I think your intent with these methods is to return correctly-shaped data rather than raising an exception, even in cases when the actual data couldn't be retrieved. While I think that makes sense for a reporting component that should try to get as much data as possible rather than crashing, I could imagine someone running the tool and being surprised to find that the results from this method are all 0, and it would be difficult to debug why with the current architecture.

I'd suggest adding some warning+-level logging so that, if a user did want to investigate why they got the stats back that they did, they'd have somewhere to start. For example, the `get_api_data()` call below could fail because the remote URL isn't in the right format, and without any reporting either in `get_api_data()` or here it'd be difficult to determine that that's the source of the issue.

d33bs replied:
Thanks, great point! I'll add in a warning message bout this.

falquaddoomi reviewed (commented):
The API you're integrating with seems neat, and your implementation here is straightforward. I had a comment about error reporting, but aside from that this looks good.

d33bs commented on src/almanack/metrics/data.py:
```diff
@@ -379,6 +379,9 @@ def compute_repo_data(repo_path: str) -> None:
     )
 
+    # gather data from ecosystems packages api
+    packages_data = get_ecosystems_package_metrics(repo_url=remote_url)
```
Thanks, good idea, I'll add in a test case for this!

d33bs commented:
Thanks @falquaddoomi for the review! I made some changes based on your comments. Additionally I decided to increase the backoff for API requests due to repeated errors experienced here and in other PR's to help address things incrementally. Merging this in now.

falquaddoomi commented on src/almanack/metrics/data.py:
```diff
@@ -770,3 +776,57 @@ def get_github_build_metrics(
+        repo_url = repo_url.replace("http://", "https://")
+
+    # perform package srequest
+    package_data = get_api_data(
+        api_endpoint="https://packages.ecosyste.ms/api/v1/packages/lookup",
+        params={"repository_url": repo_url},
+    )
```
Nice to see the warning messages you added above and the fix to translate it into HTTPS.

This call to `get_api_data()` is the one I'm mostly concerned about; if it's given a URL like `git@github.com:dangoslen/changelog-enforcer.git`, which would happen if someone cloned the repo via SSH, the API returns an empty list. It might help someone trying to debug the problem if you checked whether `package_data` is an empty list and emitted a warning like `f"Got an empty response from API for URL {repo_url}"`, in which case the user might see it was the wrong type of URL and might be able to fix it.

Alternatively, you could detect when the repo URL is `git@github.com:<user>/<repo>.git` and map it to `http://github.com/<user>/<repo>`, like how you did with the HTTP->HTTPS fix. It might be a good idea to do that anyway since I imagine this won't be an uncommon source of errors, and still have some logging when the API doesn't produce a response.

d33bs replied:
Thanks for this! I hadn't thought about the SSH / `.git` URL's! I'll update this to do as you mentioned, attempting to convert the path to an HTTPS link.

falquaddoomi reviewed (commented):
Looking good! I clarified my concerns a bit with repo URL handling in `get_ecosystems_package_metrics()`. I think once SSH repo URLs are addressed this will be good to merge; I'll leave what to log to your discretion.

d33bs commented:
Thanks @falquaddoomi for the additional comments! I've added in capability now to accept the `git@` links (abstracted to accommodate multiple platforms) and added some additional logging messages.

falquaddoomi reviewed (approved):
I see that you added a lot of logging to the `get_api_data()` function; thank you for taking my requests into consideration! The handler to remap SSH-style repo URLs looks good to me as well, and the fact that you added it to the test is great.

Nice work, and thanks again for bearing with me through all my requests. 😄

d33bs commented:
Thanks a bunch @falquaddoomi - great comments throughout! Merging this in now.
